File: 정민선/프로그래머스/[3차] 파일명 정렬.py
def solution(files):
    answer = []
    
    def section(file):
        start = -1
        
        for i in range(len(file)):
            if file[i].isdigit():
                start = i
                break
                
        head = file[:start]
        number = file[start:start+5]
        
        end = start+5
        for i in range(start, len(file)):
            if not file[i].isdigit():
                end = i
                break

        number = file[start:end]
        tail = file[end:]
        
        return head, number, tail
    
    
    fileList = []
    for file in files:
        head, number, tail = section(file)
        fileList.append((head, number, tail))
    
    fileList.sort(key=lambda x : (x[0].lower(), int(x[1])))
    
    for head, number, tail in fileList:
        answer.append(f"{head}{number}{tail}")
    
    return answer

File: 이진영/프로그래머스/[3차] 파일명 정렬.js
function solution(files) {    
    const objs = files.map((file, idx) => seperateFilename(file, idx));

    return objs.sort((a, b) => {
        const headCom = compareString(a.HEAD, b.HEAD);
        if (headCom !== 0) return headCom;
        else if (a.NUMBER !== b.NUMBER) return a.NUMBER - b.NUMBER;
        else return a.idx - b.idx;
    }).map(v => v.total);
}

// 파일명 -> HEAD, TAIL, idx, name으로 분리
const seperateFilename = (name, idx) => {
    const regex = new RegExp("^(\\D+)(\\d+)(.*)$");
    const [HEAD, NUMBER, _] = name.match(regex).slice(1, 4);
    
    return {
        HEAD,
        NUMBER: +NUMBER,
        idx: idx,
        total: name
    };
}

// 대문자, 소문자 상관없이 비교
const compareString = (a, b) => {
    const aL = a.toLowerCase();
    const bL = b.toLowerCase();
    
    if (aL > bL) return 1;
    else if (aL < bL) return -1;
    else return 0;
}